File: android/android/src/de/benibela/videlibri/NewLibrary.java
package de.benibela.videlibri;


import android.os.Bundle;
import android.sax.TextElementListener;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.*;

import java.lang.reflect.Array;
import java.util.HashMap;
import java.util.Map;

public class NewLibrary extends VideLibriBaseActivity{


    static final int MODE_LIBRARY_MODIFY = 1237;

    Bridge.LibraryDetails details;


    HashMap<String, Pair<String, EditText>> variables = new HashMap<String, Pair<String, EditText>>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);    //To change body of overridden methods use File | Settings | File Templates

        setContentView(R.layout.newlib);

        findButtonById(R.id.install).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bridge.VLInstallLibrary(getTextViewText(R.id.url));
                setLoading(true);
            }
        });

        final String [] templates = Bridge.VLGetTemplates();
        final Spinner templatesSpinner = (Spinner) findViewById(R.id.template);
        ArrayAdapter<String> adapter = new ArrayAdapter(this, android.R.layout.simple_spinner_item, templates);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        templatesSpinner.setAdapter(adapter);

        templatesSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (i < 0 || i >= templates.length) return;
                selectTemplate(templates[i]);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });


        final int mode = getIntent().getIntExtra("mode", 0);
        if (mode == MODE_LIBRARY_MODIFY) {
            final String id = getIntent().getStringExtra("libId");
            if (id == null) return;
            details = Bridge.VLGetLibraryDetails(id);
            if (details == null) return;

            for (int i=0;i<details.variableNames.length;i++)
                variables.put(details.variableNames[i], new Pair<String, EditText>(details.variableValues[i], null));
            for (int i=0;i<templates.length;i++)
                if (details.templateId.equals(templates[i])) {
                    templatesSpinner.setSelection(i);
                    break;
                }
            //selectTemplate(details.templateId);

            findViewById(R.id.createGroup).setVisibility(View.GONE);
            findButtonById(R.id.create).setText("ändern");

            setTextViewText(R.id.id, id);
            setTextViewText(R.id.name, details.prettyName);
            //(
            /*ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this,
        R.array.planets_array, android.R.layout.simple_spinner_item);
// Specify the layout to use when the list of choices appears
adapter.
// Apply the adapter to the spinner
spinner.setAdapter(adapter);*/

            findViewById(R.id.create).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    String newId = getTextViewText(R.id.id);
                    int sublocs = newId.split("_").length;
                    for (int i=sublocs; i < 4;i++) newId = "-_" + newId;

                    details.prettyName = getTextViewText(R.id.name);
                    if (newId != id) Bridge.VLSetLibraryDetails(id, null);

                    Bridge.VLSetLibraryDetails(newId, details);
                    finish();
                }
            });
            findViewById(R.id.deleteButton).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Bridge.VLSetLibraryDetails(id, null);
                    finish();
                }
            });
        } else setTextViewText(R.id.id, "user"+(int)(Math.random()*1000));

        findViewById(R.id.create).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String newId = getTextViewText(R.id.id);
                int sublocs = newId.split("_").length;
                for (int i=sublocs; i < 4;i++) newId = "-_" + newId;

                if (details == null) details = new Bridge.LibraryDetails();

                details.prettyName = getTextViewText(R.id.name);
                if (mode == MODE_LIBRARY_MODIFY && getIntent().getStringExtra("libId") != null) {
                    String id = getIntent().getStringExtra("libId");
                    if (!newId.equals(id)) Bridge.VLSetLibraryDetails(id, null);
                }
                details.id = newId;
                details.templateId = templatesSpinner.getSelectedItem().toString();

                details.variableNames = new String[variables.size()];
                details.variableValues = new String[variables.size()];
                int i = 0;
                for (Map.Entry<String, Pair<String, EditText>> e: variables.entrySet()){
                    details.variableNames[i] = e.getKey();
                    if (e.getValue().second == null) details.variableValues[i] = e.getValue().first;
                    else details.variableValues[i] = e.getValue().second.getText().toString();
                    i+=1;
                }


                Bridge.VLSetLibraryDetails(newId, details);
                finish();
            }
        });

    }

    EditText addTemplateVariable(LinearLayout linearLayout, LayoutInflater inflater, String name, String desc, String value){
        View option = inflater.inflate(R.layout.newliboption, null);
        TextView text = (TextView) option.findViewById(R.id.text);
        text.setText(desc == null ? name : (name +  " ( "+desc+")"));
        EditText edit = (EditText) option.findViewById(R.id.edit);
        if (value == null) value = "";
        edit.setText(value);
        linearLayout.addView(option);
        variables.put(name, new Pair<String, EditText>(value, edit));
        return edit;
    }

    void selectTemplate(String template){
        Bridge.TemplateDetails details = Bridge.VLGetTemplateDetails(template);
        if (details == null) return;

        HashMap<String, Pair<String, EditText>> oldVariables = variables;
        variables = new HashMap<String, Pair<String, EditText>>();

        HashMap<String, String> oldValues = new HashMap<String, String>();

        for (Map.Entry<String, Pair<String, EditText>> e: oldVariables.entrySet() )
            if (e.getValue().second == null) oldValues.put(e.getKey(), e.getValue().first);
            else if (!e.getValue().second.getText().toString().equals(e.getValue().first))
                oldValues.put(e.getKey(), e.getValue().second.getText().toString());

        LinearLayout linearLayout = (LinearLayout) findViewById(R.id.options);
        linearLayout.removeAllViews();

        LayoutInflater inflater = getLayoutInflater();
        for (int i=0;i<details.variablesNames.length;i++) {
            EditText et = addTemplateVariable(    linearLayout, inflater, details.variablesNames[i], details.variablesDescription[i], details.variablesDefault[i]);
            if (oldValues.containsKey(details.variablesNames[i])) et.setText(oldValues.get(details.variablesNames[i]));
        }

        for (Map.Entry<String, String> e: oldValues.entrySet() )
            if (!variables.containsKey(e.getKey()))
                addTemplateVariable(linearLayout, inflater, e.getKey(), null, "").setText(e.getValue());
    }
}

File: android/android/src/de/benibela/videlibri/activities/NewLibrary.kt
package de.benibela.videlibri.activities

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.EditText
import android.widget.LinearLayout
import de.benibela.videlibri.R
import de.benibela.videlibri.databinding.NewlibBinding
import de.benibela.videlibri.databinding.NewliboptionBinding
import de.benibela.videlibri.jni.Bridge
import de.benibela.videlibri.jni.LibraryDetails
import de.benibela.videlibri.utils.*

data class LibraryVariable(val defaultValue: String, val editText: EditText?)

class NewLibrary : VideLibriBaseActivity() {
    private lateinit var binding: NewlibBinding

    var details: LibraryDetails? = null
    private var trChooseATemplate: String? = null
    private var variables = HashMap<String, LibraryVariable>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = setVideLibriView(NewlibBinding::inflate)
        trChooseATemplate = getString(R.string.lay_newlib_list_choosesystem)
        val templates = Bridge.VLGetTemplates().toMutableList().run { add(0, trChooseATemplate); toTypedArray() }
        binding.templateSpinner.setItems(templates)
        binding.templateSpinner.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                templates.getOrNull(i)?.let(::selectTemplate)
            }
            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
        binding.checkBoxSetId.setOnCheckedChangeListener { _, checked -> binding.idLayout.isVisibleNotGone = checked }
        val mode = intent.getIntExtra("mode", 0)
        if (mode == MODE_LIBRARY_MODIFY) {
            val id = intent.getStringExtra("libId") ?: return
            details = Bridge.VLGetLibraryDetails(id)
            val details = details ?: return
            details.variables.forEach { variables[it.name] = LibraryVariable(it.value, null) }
            val templatePos = templates.indexOf(details.templateId)
            if (templatePos >= 0) binding.templateSpinner.setSelection(templatePos)

            binding.create.setText(R.string.change)
            viewId = id
            binding.name.setText(details.prettyName)
            binding.deleteButton.setOnClickListener {
                showMessageYesNo(getString(R.string.delete_library_confirmation, binding.name.text.toString())) {
                    Bridge.VLSetLibraryDetails(id, null)
                    withActivity<NewLibrary> {  finishWithResult() }
                }
            }
        } else {
            viewId = ""
            binding.deleteButton.visibility = View.GONE
        }
        binding.create.setOnClickListener {
            val templateId = binding.templateSpinner.selectedItem.toString()
            if (templateId == trChooseATemplate) showMessage(R.string.lay_newlib_error_choosesystem)
            else if (variables.isNotEmpty() && variables.all { it.value.editText?.text?.isBlank() ?: true }) showMessage(R.string.lay_newlib_error_no_parameters)
            else {
                val oldId = if (mode == MODE_LIBRARY_MODIFY) intent.getStringExtra("libId") else null
                val newId = viewId
                val details = this.details ?: LibraryDetails()
                details.prettyName = binding.name.text.toString()
                details.id = newId
                details.templateId = templateId
                details.variables = variables.map { (key, value) -> de.benibela.videlibri.jni.LibraryVariable(key, value.editText?.text?.toString() ?: value.defaultValue) }.toTypedArray()
                Bridge.VLSetLibraryDetails(newId, details)
                if (newId != oldId && oldId != null) Bridge.VLSetLibraryDetails(oldId, null)
                finishWithResult()
            }
        }
        //binding.name.requestFocus() //or use scrollView.requestChildFocus(target, target); ??
    }

    private var viewId
        get() = arrayOf(binding.idCountry, binding.idState, binding.idCity, binding.idId).joinToString("_") {
            it.text.toString().replace('_', '-').trim().takeNonEmpty() ?: "-"
        }
        set(value){
            val sid = value.split("_").toMutableList()
            if (sid.size > 4) sid[3] = sid.drop(3).joinToString("_")
            while (sid.size < 4) sid.add(0, "")
            if (sid[3].isEmpty()) {
                val existingUserLibraries = Bridge.VLGetOptions().userLibIds
                sid[3] = if (existingUserLibraries.isEmpty()) "user"
                else run {
                    val oldIds = existingUserLibraries.map { it.substringAfterLast("user").toIntOrNull() ?: 0 }
                    "user${(oldIds.maxOrNull() ?: existingUserLibraries.size) + 1}"
                }
            }
            arrayOf(binding.idCountry, binding.idState, binding.idCity, binding.idId).zip(sid).forEach { it.first.setText(it.second) }
        }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        binding.idLayout.isVisibleNotGone = binding.checkBoxSetId.isChecked
    }

    private fun addTemplateVariable(linearLayout: LinearLayout, inflater: LayoutInflater, name: String, desc: String?, defaultValue: String?): EditText {
        val value = defaultValue ?: ""
        val option = NewliboptionBinding.inflate(inflater, linearLayout, true)
        option.text.text = desc?.let { "$name ( $desc)" } ?: name
        option.edit.setText(value)
        variables[name] = LibraryVariable(value, option.edit)
        return option.edit
    }

    fun selectTemplate(template: String) {
        if (template == trChooseATemplate) return
        val details =
            try {
                Bridge.VLGetTemplateDetails(template) ?: return
            } catch (e: de.benibela.videlibri.jni.InternalError) {
                showMessage(e.localizedMessage)
                return
            }

        binding.textViewTemplateDescription.text = details.description
        val oldVariables = variables
        variables = HashMap()
        val oldValues = HashMap<String, String>()
        for ((key, value) in oldVariables)
            if (value.editText == null) oldValues[key] = value.defaultValue
            else if (value.editText.text.toString() != value.defaultValue) oldValues[key] = value.editText.text.toString()
        val linearLayout = binding.options
        linearLayout.removeAllViews()
        val inflater = layoutInflater
        for (i in details.variablesNames.indices) {
            val et = addTemplateVariable(linearLayout, inflater, details.variablesNames[i], details.variablesDescription[i], details.variablesDefault[i])
            if (details.variablesNames[i] in oldValues) et.setText(oldValues[details.variablesNames[i]])
        }
        for ((key, value) in oldValues) if (key !in variables) addTemplateVariable(linearLayout, inflater, key, null, "").setText(value)
    }

    companion object {
        const val MODE_LIBRARY_MODIFY = 1237
    }
}